src/eddymotion/model/kernels.py

```diff
@@ -0,0 +1,190 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+import numpy as np
+from sklearn.gaussian_process.kernels import Hyperparameter, Kernel
+
+from eddymotion.model.utils import compute_pairwise_angles
+
+
+class SphericalCovarianceKernel(Kernel):
+    """
+    Custom kernel based on spherical covariance function.
+
+    Parameters
+    ----------
+    lambda_s : :obj:`float`
+        Scale parameter for the covariance function.
+    a : :obj:`float`
+        Distance parameter where the covariance function goes to zero.
+    sigma_sq : :obj:`float`
+        Noise variance term.
+    lambda_s_bounds : :obj:`tuple` of :obj:`float` or "fixed"
+        The lower and upper bound on lambda_s.
+    a_bounds : :obj:`tuple` of :obj:`float` or "fixed"
+        The lower and upper bound on a.
+    sigma_sq_bounds : :obj:`tuple` of :obj:`float` or "fixed"
+        The lower and upper bound on sigma_sq.
+    """
+
+    def __init__(
+        self,
+        lambda_s=2.0,
+        a=0.1,
+        sigma_sq=1.0,
+        lambda_s_bounds=(1e-5, 1e4),
+        a_bounds=(1e-5, np.pi),
+        sigma_sq_bounds=(1e-5, 1e4),
+    ):
+        self.lambda_s = lambda_s
+        self.a = a
+        self.sigma_sq = sigma_sq
+        self.lambda_s_bounds = lambda_s_bounds
+        self.a_bounds = a_bounds
+        self.sigma_sq_bounds = sigma_sq_bounds
+
+    @property
+    def hyperparameter_lambda_s(self):
+        return Hyperparameter("lambda_s", "numeric", self.lambda_s_bounds)
+
+    @property
+    def hyperparameter_a(self):
+        return Hyperparameter("a", "numeric", self.a_bounds)
+
+    @property
+    def hyperparameter_sigma_sq(self):
+        return Hyperparameter("sigma_sq", "numeric", self.sigma_sq_bounds)
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """
+        Compute the kernel matrix.
+
+        Parameters
+        ----------
+        X : :obj:`~numpy.ndarray`
+            Diffusion gradient encoding directions in FSL format.
+        Y : :obj:`~numpy.ndarray`
+            Diffusion gradient encoding directions in FSL format., optional
+        eval_gradient : :obj:`bool`
+            Evaluate the gradient with respect to the log of
+            the kernel hyperparameter.
+
+        Returns
+        -------
+        K : :obj:`~numpy.ndarray` of shape (n_samples, n_samples)
+            Kernel matrix.
+        K_gradient : :obj:`~numpy.ndarray` of shape (n_samples, n_samples, n_dims), optional
+            The gradient of the kernel matrix with respect to the log of the
+            hyperparameters. Only returned when `eval_gradient` is True.
+        """
+        if Y is not None:
+            angles = compute_pairwise_angles(X.T, Y.T, closest_polarity=True)
+            K = self.lambda_s * np.where(
```

Comment:
We probably want to call this
https://github.com/nipreps/eddymotion/pull/213/files#diff-ba53cdfdd502439a31a24c148f9be58dab5fb56d09b25547557933b3b259e46fR57

Making that a function will allow for reusing it in the multi-shell case. Also, will allow us to ensure that the behavior is the expected one prior to the kernel computations (e.g. together with PR #214).